dipy/direction/ptt_direction_getter.pyx

```diff
@@ -0,0 +1,472 @@
+
+        cdef double data_support = 0
+        cdef double max_posterior = 0
+
+        cdef int tries
+        for tries in range(1000):
+            self.initialize_candidate(seed_direction)
+            data_support = self.calculate_data_support()
+            if data_support > max_posterior:
+                max_posterior = data_support
+
+        # Compensation for underestimation of max posterior estimate
+        max_posterior = pow(2.0 * max_posterior, self.data_support_exponent)
+
+        # Initialization is successful if a suitable candidate can be sampled
+        # within the trial limit
+        for tries in range(1000):
+            self.initialize_candidate(seed_direction)
+            if (random() * max_posterior <= self.calculate_data_support()):
+                self.last_val = self.last_val_cand
+                return 0
+        return 1
+
+    cdef int propagate(self):
+        """Propagates the position by step_size amount. The propagation is using
+        the parameters of the last candidate curve. Then, randomly generate
+        curve parametrization from the current position. The walking frame
+        is the same, only the k1 and k2 parameters are randomly picked.
+        Rejection sampling is used to pick the next curve using the data
+        support (likelihood).
+
+        Returns
+        -------
+        status : int
+            Returns 0 if the propagation was successful, or
+            1 otherwise.
+        """
+        cdef double max_posterior = 0
+        cdef double data_support = 0
+        cdef double[3] tangent
+        cdef int tries = 0
+
+        self.prepare_propagator(self.step_size)
+
+        # propagate
+        for i in range(3):
+            self.position[i] = (self.propagator[0] * self.frame[0][i] +
+                                self.propagator[1] * self.frame[1][i] +
+                                self.propagator[2] * self.frame[2][i] +
+                                self.position[i])
+            tangent[i] = (self.propagator[3] * self.frame[0][i]
+                          + self.propagator[4] * self.frame[1][i]
+                          + self.propagator[5] * self.frame[2][i])
+            self.frame[2][i] = (self.propagator[6] * self.frame[0][i] +
+                                self.propagator[7] * self.frame[1][i] +
+                                self.propagator[8] * self.frame[2][i])
+        normalize(tangent)
+        cross(self.frame[1], self.frame[2], tangent)
+        normalize(self.frame[1])
+        cross(self.frame[2], tangent, self.frame[1])
+        self.frame[0][0] = tangent[0]
+        self.frame[0][1] = tangent[1]
+        self.frame[0][2] = tangent[2]
+
+        # This is adaptively set in Trekker.
+        for tries in range(20):
+            self.k1, self.k2 = random_point_within_circle(self.max_curvature)
+            data_support = self.calculate_data_support()
+            if data_support > max_posterior:
+                max_posterior = data_support
```

Comment:
same questions as above

Reply:
I added the variable and reduced the value to 10, which almost 2X the speed of PTT. In my tests, the results are not different. We need optimizing this in a future PR.

Reply:
If speed is a big concern 10 will work too, although in that case max_posterior will be underestimated quite often.

Reply:
ok. Currently, tries for suitable propagation direction is set to 10. The max posterior estimation is set to 100 samples.